server/rest/src/main/java/org/infinispan/rest/InfinispanRequestCreator.java

```diff
@@ -12,27 +18,56 @@
  */
 class InfinispanRequestCreator {
 
+   private static final String SEARCH_ACTION = "search";
+   private static final String ACTION_PARAMETER = "action";
+
    /**
     * Creates the appropriate {@link InfinispanRequest} instance based on the raw incoming request.
     */
    static InfinispanRequest createRequest(FullHttpRequest request, ChannelHandlerContext ctx) {
       QueryStringDecoder queryStringDecoder = new QueryStringDecoder(request.uri());
+      Map<String, List<String>> parameters = queryStringDecoder.parameters();
+
+      String[] components = queryStringDecoder.path().substring(1).split("/");
+
+      String context;
+      Optional<String> key = Optional.empty();
+      Optional<String> cacheName = Optional.empty();
+
+      List<String> actionParameter = parameters.get(ACTION_PARAMETER);
+
+      if (components.length > 3 || components.length == 0) {
+         throw new MalformedRequest("Invalid request path");
+      }
+      Iterator<String> pathElements = Arrays.stream(components).iterator();
+      context = pathElements.next();
+      if (pathElements.hasNext()) {
+         cacheName = Optional.of(pathElements.next());
+      }
+      if (pathElements.hasNext()) {
+         key = Optional.of(pathElements.next());
+      }
 
-      StringTokenizer pathTokenizer = new StringTokenizer(queryStringDecoder.path(), "/");
-      String context = pathTokenizer.nextToken();
-      Optional<String> cacheName, key;
-      if (pathTokenizer.hasMoreTokens()) {
-         String nextToken = pathTokenizer.nextToken();
-         cacheName = Optional.of(nextToken);
-      } else {
-         cacheName = Optional.empty();
+      if (actionParameter == null || actionParameter.isEmpty()) {
+         // Cache API request
+         return new InfinispanCacheAPIRequest(request, ctx, cacheName, key, context, parameters);
       }
-      if (pathTokenizer.hasMoreTokens()) {
-         String next = pathTokenizer.nextToken();
-         key = Optional.of(next);
-      } else {
-         key = Optional.empty();
+
+      if (actionParameter.size() > 1) {
+         throw new MalformedRequest("The 'action' parameter must contain only one value");
       }
-      return new InfinispanCacheAPIRequest(request, ctx, cacheName, key, context);
+
+      String action = actionParameter.iterator().next();
+
+      switch (action) {
```

Comment:
Switch with one case? :)

Reply:
Soon we'll have counters, scripts, etc, so I've already provided the switch :)